genie-agent/src/main/java/com/netflix/genie/agent/execution/services/AgentJobSpecificationService.java

```diff
@@ -0,0 +1,50 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.execution.services;
+
+import com.netflix.genie.common.dto.v4.JobRequest;
```

Comment:
Do you expect this DTO to work for both the `v4` API to submit jobs AND this internal job request resolution?

I haven't given much thought to wether that is the case or not, but in general seems strange to use a public API entity for an internal protocol.
For one thing we may want to evolve them separately.
Just a gut feeling that they're going to diverge more and more.

Example: Container information, it applies to the API but not here.
Maybe the API job request is a superset of the internal fields?

Reply:
I do.

Firstly the Job Request is not the wire level protocol between the two components and it is translated on both sides by the adapter so it's already shielded there.

Second, the job request as it currently stands consists of only 4 required fields (`name`, `user`, `clusterCriteria`, `commandCriterion`). Everything else is completely optional.

Thirdly we can extend `JobRequest` or wrap it in a larger payload representation for an API request (which I believe is what you mean by your last line in your comment) e.g.
```
POST /api/v4/jobs
{
   "additionalInfo" : { ... },
   "jobRequest": { ... },
   "containerParameters": { ... }
}
```

I don't have much interest in creating yet another data class (server side dto, wire protocol proto file, agent side job metadata something) which will end up representing the same information. My feeling is that regardless of job launching path there are certain fields that are necessary for Genie to properly handle a job and I've tried to capture these in this `JobRequest` DTO.

Reply:
Additionally I definitely view both agent initiated and REST API initiated calls going to the `JobSpecificationServiceImpl` APIs on the server. I view the REST API path looking something like `API -> JobOrchestrator -> JobSpecificationServiceImpl` and inside whatever form that `JobOrchestrator` takes we'll be persisting things to the database (like the initial job metadata: `id`, `name`, `user`, etc). I also think the `resolveJobSpecification` API within this implementation will take care of persisting which cluster, command, etc was chosen for the job.